### PR TITLE
[NLL] Better move errors

### LINKED ISSUE #45699: MIR borrowck: diagnostic for moves out of patterns should be improved
When a `match` expression or `let` statement is used on an lvalue with no ability to move things out, AST borrowck displays a single unified error message while MIR borrowck displays an error message for each binding:
```Rust
#![feature(box_syntax)]

enum Foo {
    Foo1(Box<u32>, Box<u32>),
    Foo2(Box<u32>),
    Foo3,
}

fn blah() {
    let f = &Foo::Foo1(box 1, box 2);
    match *f {             //~ ERROR cannot move out of
                           //~| cannot move out
        Foo::Foo1(_num1,         //~ NOTE to prevent move
                  _num2) => (),  //~ NOTE and here
        Foo::Foo2(_num) => (),   //~ NOTE and here
        Foo::Foo3 => ()
    }
}

fn main() {}
```

Displays the following errors:
```
error[E0507]: cannot move out of borrowed content (Ast)
  --> x.rs:11:11
   |
11 |     match *f {             //~ ERROR cannot move out of
   |           ^^ cannot move out of borrowed content
12 |                            //~| cannot move out
13 |         Foo::Foo1(_num1,         //~ NOTE to prevent move
   |                   ----- hint: to prevent move, use `ref _num1` or `ref mut _num1`
14 |                   _num2) => (),  //~ NOTE and here
   |                   ----- ...and here (use `ref _num2` or `ref mut _num2`)
15 |         Foo::Foo2(_num) => (),   //~ NOTE and here
   |                   ---- ...and here (use `ref _num` or `ref mut _num`)

error[E0507]: cannot move out of borrowed_content (Mir)
  --> x.rs:13:19
   |
13 |         Foo::Foo1(_num1,         //~ NOTE to prevent move
   |                   ^^^^^ cannot move out of borrowed_content

error[E0507]: cannot move out of borrowed_content (Mir)
  --> x.rs:14:19
   |
14 |                   _num2) => (),  //~ NOTE and here
   |                   ^^^^^ cannot move out of borrowed_content

error[E0507]: cannot move out of borrowed_content (Mir)
  --> x.rs:15:19
   |
15 |         Foo::Foo2(_num) => (),   //~ NOTE and here
   |                   ^^^^ cannot move out of borrowed_content

error: aborting due to 4 previous errors
```

I think this is fairly low priority, but we do want error parity.

### LINKED ISSUE #46627: MIR borrowck: missing suggestion to use a reference
When moving out of a value into a new let binding, AST borrowck suggests using a reference, but MIR borrowck does not.

e.g. for E0508:
```Rust
struct NonCopy;

fn main() {
    let array = [NonCopy; 1];
    let _value = array[0];  //[ast]~ ERROR [E0508]
                            //[mir]~^ ERROR [E0508]
}
```

We have the following error message
```
error[E0508]: cannot move out of type `[NonCopy; 1]`, a non-copy array (Ast)
  --> /home/ariel/Rust/rust-master/src/test/compile-fail/E0508.rs:18:18
   |
18 |     let _value = array[0];  //[ast]~ ERROR [E0508]
   |                  ^^^^^^^^
   |                  |
   |                  cannot move out of here
   |                  help: consider using a reference instead: `&array[0]`

error[E0508]: cannot move out of type `[NonCopy; 1]`, a non-copy array (Mir)
  --> /home/ariel/Rust/rust-master/src/test/compile-fail/E0508.rs:18:18
   |
18 |     let _value = array[0];  //[ast]~ ERROR [E0508]
   |                  ^^^^^^^^ cannot move out of here

error: aborting due to 2 previous errors
```

See that the help is only in the AST. We should port that help message to MIR borrowck. 

The help message is generated in this place in AST borrowck:
https://github.com/rust-lang/rust/blob/02b4d3ddfb7cd9f1caf8ed90f18f8e798a1f4b05/src/librustc_borrowck/borrowck/gather_loans/move_error.rs#L77-L87

### LINKED ISSUE #51187: NLL: missing suggestions to "consider using a reference instead"
AST borrowck suggests using a reference when you try to move something that cannot be moved (e.g. out of an array, or out of borrowed content, or out of a struct that implements `Drop`).

NLL fails to provide this suggestion; it just issues an error.

Here are the test cases where pnkfelix has noted this:

 * [ui/E0508.ast.nll.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/E0508.ast.nll.stderr); compare with [E0508.ast.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/E0508.ast.stderr)
 * [ui/error-codes/E0509.nll.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/error-codes/E0509.nll.stderr); compare with [E0509.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/error-codes/E0509.stderr)
 * [ui/issue-40402-ref-hints/issue-40402-1.nll.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/issue-40402-ref-hints/issue-40402-1.nll.stderr); compare with [issue-40402-1.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/issue-40402-ref-hints/issue-40402-1.stderr)

### LINKED ISSUE #51189: NLL fails to suggest using `ref` when a pattern is otherwise an impossible move
In the case of [ui/borrowck/borrowck-move-error-with-note.rs](https://github.com/rust-lang/rust/blob/master/src/test/ui/borrowck/borrowck-move-error-with-note.rs), the AST borrowck output ([borrowck-move-error-with-note.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/borrowck/borrowck-move-error-with-note.stderr)) includes suggestions like this:

```
LL |         Foo::Foo1(num1,
   |                   ---- hint: to prevent move, use `ref num1` or `ref mut num1`

```

but NLL ([borrowck-move-error-with-note.nll.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/borrowck/borrowck-move-error-with-note.nll.stderr)) presents no such suggestion for errors it discovers in the MIR for `match` patterns.

Here are all the tests that I think involve this:
 * [ ] [ui/borrowck/borrowck-move-error-with-note.nll.stderr](https://github.com/rust-lang/rust/blob/master/src/test/ui/borrowck/borrowck-move-error-with-note.nll.stderr)
 * [ ] [ui/borrowck/borrowck-vec-pattern-nesting.nll.stderr](https://github.com/rust-lang/rust./blob/master/src/test/ui/borrowck/borrowck-vec-pattern-nesting.nll.stderr)

### PR DESCRIPTION
**Pull request by matthewjasper**
Make a number of changes to improve the quality of NLL cannot move errors.

* Group errors that occur in the same `match` with the same cause.
* Suggest `ref`, `&` or removing `*` to avoid the move.
* Show the place being matched on.

Differences from AST borrowck:

* `&` is suggested over `ref` when matching on a place that can't be moved from.
* Removing `*` is suggested instead of adding `&` when applicable.
* Sub-pattern spans aren't used, this would probably need Spans on Places.

Closes #45699
Closes #46627
Closes #51187
Closes #51189

r? @pnkfelix

### CODE REVIEW

File: src/librustc_mir/build/matches/mod.rs [hunk 1]
```diff
@@ -288,6 +298,20 @@ impl<'a, 'gcx, 'tcx> Builder<'a, 'gcx, 'tcx> {
                       candidate.match_pairs);
         }
 
+        if let Some(span) = initializer_span {
```
Comment (nikomatsakis) [inline src/librustc_mir/build/matches/mod.rs]: it feels to me like a comment would be nice here

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 2]
```diff
@@ -0,0 +1,364 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: Thank you for making a module <3

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+
+#[derive(Copy, Clone)]
+struct MoveErrorCtxt<'a, 'gcx: 'tcx, 'tcx: 'a> {
+    mir: &'a Mir<'tcx>,
+    tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    move_data: &'a MoveData<'tcx>,
+}
+
+#[derive(Debug)]
+enum GroupedMoveError<'tcx> {
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: I feel like a meta comment on the enum explaining (in general) what we group by and why would be super useful. e.g.

---

often when desugaring a pattern match we may have many individual moves in MIR
that are all part of one operation from the user's point-of-view. For example:

```
let (x, y) = foo()
```

would move `x` from the `0` field of some temporary, and `y` from the `1` field. We group such errors together for cleaner error reporting.

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    // Match place can't be moved from
+    // e.g. Matching on x[0].
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: can you make this rust code? I guess this means `match x[0] { ... }` where `x` has type `&[u32]` or something?

Comment (nikomatsakis) [reply src/librustc_mir/borrow_check/move_errors.rs]: (Of course the same is true for `&Vec`...)

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+    MovesFromMatchPlace {
+        span: Span,
+        move_from: Place<'tcx>,
+        kind: IllegalMoveOriginKind<'tcx>,
+        binds_to: Vec<Local>,
+    },
+    // Part of a pattern can't be moved from,
+    // e.g. moving from a &x pattern.
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: similarly, I would say e.g. `let &x = ...` -- here `x` references borrowed content, so a move would be illegal

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+                    .map(|stmt| &stmt.kind)
+                {
+                    let local_decl = &self.mir.local_decls[*local];
+                    if let Some(ClearCrossCrate::Set(BindingForm::Var(VarBindingForm {
+                        ref opt_match_place,
+                        binding_mode: _,
+                        opt_ty_info: _,
+                    }))) = local_decl.is_user_variable
+                    {
+                        let (match_place, match_span) = opt_match_place
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: I'd like to see some sort of comment explaining what these represent in terms of Rust source (e.g., give a Rust example, and explain what `match_place` and `match_span` would be)

File: src/librustc_mir/borrow_check/move_errors.rs [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+                            .as_ref()
+                            .map(|&(ref x, y)| (x, y))
+                            .unwrap_or((move_from, stmt_source_info.span));
+
+                        if self
```
Comment (nikomatsakis) [inline src/librustc_mir/borrow_check/move_errors.rs]: I don't quite get why this `sub_scope` test makes sense... example of where it applies (and maybe where it doesn't?)

**Review by nikomatsakis — APPROVED** [review]
Left a bunch of nits, but this is really quite nice. 💯

**Comment by matthewjasper** [comment]
Nits addressed

**Review by nikomatsakis — APPROVED** [review]

File: src/test/mir-opt/inline-closure.rs [hunk 8]
```diff
@@ -36,7 +36,7 @@ fn foo<T: Copy>(_t: T, q: i32) -> i32 {
 //     _5 = (move _6, move _7);
 //     _8 = move (_5.0: i32);
 //     _9 = move (_5.1: i32);
-//     _0 = move _8;
+//     _0 = _8;
```
Comment (eddyb) [inline src/test/mir-opt/inline-closure.rs]: I am very confused as to why this happened. Why was it a move before and not after?

**Pull request merged**